### PR TITLE
chore: make client-side instrumentation non-nullable

### PR DESCRIPTION
**Pull request by pavelfeldman**

### CODE REVIEW

File: packages/playwright-core/src/client/fetch.ts [hunk 1]
```diff
@@ -80,7 +77,7 @@ export class APIRequest implements api.APIRequest {
     this._contexts.add(context);
     context._request = this;
     context._tracing._tracesDir = tracesDir;
-    await this._onDidCreateContext?.(context);
+    await this._playwright._instrumentation.onDidCreateRequestContext(context);
```
Comment (dgozman) [inline packages/playwright-core/src/client/fetch.ts:80]: To be similar to other places:
```suggestion
    await this._instrumentation.onDidCreateRequestContext(context);
```

File: packages/playwright-core/src/client/page.ts [hunk 2]
```diff
@@ -691,6 +691,9 @@ export class Page extends ChannelOwner<channels.PageChannel> implements api.Page
   async pause() {
     if (require('inspector').url())
       return;
+    this._browserContext.setDefaultNavigationTimeout(0);
+    this._browserContext.setDefaultTimeout(0);
```
Comment (dgozman) [inline packages/playwright-core/src/client/page.ts:695]: I wonder if we should restore the timeout back, now that this code triggers with just a library?

File: packages/playwright-test/src/index.ts [hunk 3]
```diff
@@ -361,46 +370,26 @@ const playwrightFixtures: Fixtures<TestFixtures, WorkerFixtures> = ({
...
-      (browserType as any)._onWillCloseContext = onWillCloseContext;
       (browserType as any)._defaultContextOptions = _combinedContextOptions;
       const promises: Promise<void>[] = [];
       const existingContexts = Array.from((browserType as any)._contexts) as BrowserContext[];
       for (const context of existingContexts) {
         if ((context as any)[kIsReusedContext])
           reusedContexts.add(context);
         else
-          promises.push(onDidCreateBrowserContext(context));
+          promises.push((playwright as any)._instrumentation.onDidCreateBrowserContext(context));
```
Comment (dgozman) [inline packages/playwright-test/src/index.ts:384]:
```suggestion
          promises.push(csiListener.onDidCreateBrowserContext(context));
```

File: packages/playwright-test/src/index.ts [hunk 4]
```diff
@@ -361,46 +370,26 @@ const playwrightFixtures: Fixtures<TestFixtures, WorkerFixtures> = ({
...
       await Promise.all(promises);
     }
     {
-      (playwright.request as any)._onDidCreateContext = onDidCreateRequestContext;
-      (playwright.request as any)._onWillCloseContext = onWillCloseRequestContext;
       (playwright.request as any)._defaultContextOptions = { ..._combinedContextOptions };
       (playwright.request as any)._defaultContextOptions.tracesDir = path.join(_artifactsDir(), 'traces');
       const existingApiRequests: APIRequestContext[] =  Array.from((playwright.request as any)._contexts as Set<APIRequestContext>);
-      await Promise.all(existingApiRequests.map(onDidCreateRequestContext));
+      await Promise.all(existingApiRequests.map(c => (playwright as any)._instrumentation.onDidCreateRequestContext(c)));
```
Comment (dgozman) [inline packages/playwright-test/src/index.ts:392]:
```suggestion
      await Promise.all(existingApiRequests.map(c => csiListener.onDidCreateRequestContext(c)));
```

File: packages/playwright-test/src/index.ts [hunk 5]
```diff
@@ -361,46 +370,26 @@ const playwrightFixtures: Fixtures<TestFixtures, WorkerFixtures> = ({
...
-      await stopTracing(tracing, (context as any)[kStartedContextTearDown]);
-    };
-
     // 1. Setup instrumentation and process existing contexts.
+    const instrumentation = (playwright as any)._instrumentation as ClientInstrumentation;
+    instrumentation.addListener(csiListener);
     for (const browserType of [playwright.chromium, playwright.firefox, playwright.webkit]) {
-      (browserType as any)._onDidCreateContext = onDidCreateBrowserContext;
-      (browserType as any)._onWillCloseContext = onWillCloseContext;
       (browserType as any)._defaultContextOptions = _combinedContextOptions;
```
Comment (dgozman) [inline packages/playwright-test/src/index.ts:377]: Perhaps we can migrate this one as well, to something like `onWillCreateContext` or explicit `modifyNewContextOptions`?

File: packages/playwright-core/src/client/clientInstrumentation.ts [hunk 6]
```diff
@@ -15,17 +15,30 @@
...
+  onWillCloseBrowserContext(context: BrowserContext): Promise<void>;
+  onWillCloseRequestContext(context: APIRequestContext): Promise<void>;
 }
 
 export interface ClientInstrumentationListener {
-  onApiCallBegin?(apiCall: string, stackTrace: ParsedStackTrace | null, wallTime: number, userData: any): any;
-  onApiCallEnd?(userData: any, error?: Error): any;
+  onApiCallBegin?(apiCall: string, stackTrace: ParsedStackTrace | null, wallTime: number, userData: any): void;
+  onApiCallEnd?(userData: any, error?: Error): void;
+  onDidCreateBrowserContext?(context: BrowserContext): Promise<void>;
```
Comment (dgozman) [inline packages/playwright-core/src/client/clientInstrumentation.ts:37]: I think the PR title is now a bit off 😄

**Review by dgozman — APPROVED** [review]

File: packages/playwright-core/src/client/fetch.ts (hunk 1 above)
Comment (pavelfeldman) [reply packages/playwright-core/src/client/fetch.ts:80]: This unfortunately won't compile. You probably mean context._instrumentation. But it won't be similar.

File: packages/playwright-core/src/client/page.ts (hunk 2 above)
Comment (pavelfeldman) [reply packages/playwright-core/src/client/page.ts:695]: We could, we don't have means to doing that though, also does not seem to be important?

File: packages/playwright-test/src/index.ts (hunk 5 above)
Comment (pavelfeldman) [reply packages/playwright-test/src/index.ts:377]: I wanted to follow up

File: packages/playwright-core/src/client/clientInstrumentation.ts (hunk 6 above)
Comment (pavelfeldman) [reply packages/playwright-core/src/client/clientInstrumentation.ts:37]: yeah

**Pull request closed without merging**